File: tests/wasm/test_linearreg_slope.js
/**
 * WASM binding tests for LINEARREG_SLOPE indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --target nodejs -- --features wasm" first');
        throw error;
    }

    testData = loadTestData();
});

test('LINEARREG_SLOPE partial params', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.linearreg_slope_js(close, 14);
    assert.strictEqual(result.length, close.length);
});

test('LINEARREG_SLOPE accuracy', () => {
    
    const inputData = new Float64Array([100.0, 98.0, 95.0, 90.0, 85.0, 80.0, 78.0, 77.0, 79.0, 81.0]);
    
    const result = wasm.linearreg_slope_js(inputData, 5);
    
    assert.strictEqual(result.length, inputData.length);
    
    
    for (let i = 0; i < 4; i++) {
        assert(isNaN(result[i]), `Expected NaN at index ${i}, got ${result[i]}`);
    }
    
    
    const expectedValues = [-3.8, -4.6, -4.4, -3.3, -1.5, 0.3];
    
    for (let i = 0; i < expectedValues.length; i++) {
        const idx = i + 4;
        assertClose(result[idx], expectedValues[i], 1e-9, 
                   `LINEARREG_SLOPE value mismatch at index ${idx}`);
    }
});

test('LINEARREG_SLOPE zero period', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.linearreg_slope_js(inputData, 0);
    }, /Invalid period/);
});

test('LINEARREG_SLOPE period exceeds length', () => {
    
    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.linearreg_slope_js(dataSmall, 10);
    }, /Invalid period/);
});

test('LINEARREG_SLOPE very small dataset', () => {
    
    const singlePoint = new Float64Array([42.0]);
    
    assert.throws(() => {
        wasm.linearreg_slope_js(singlePoint, 14);
    }, /Invalid period/);
});

test('LINEARREG_SLOPE empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.linearreg_slope_js(empty, 14);
    }, /Empty data provided/);
});

test('LINEARREG_SLOPE fast API', () => {
    
    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
    const len = data.length;
    
    
    const inPtr = wasm.linearreg_slope_alloc(len);
    const outPtr = wasm.linearreg_slope_alloc(len);
    
    try {
        
        const inView = new Float64Array(wasm.__wasm.memory.buffer, inPtr, len);
        inView.set(data);
        
        
        wasm.linearreg_slope_into(inPtr, outPtr, len, 3);
        
        
        const result = new Float64Array(wasm.__wasm.memory.buffer, outPtr, len);
        
        
        assert(isNaN(result[0]), 'Expected NaN at index 0');
        assert(isNaN(result[1]), 'Expected NaN at index 1');
        
        
        assert(!isNaN(result[2]), 'Expected valid value at index 2');
    } finally {
        
        wasm.linearreg_slope_free(inPtr, len);
        wasm.linearreg_slope_free(outPtr, len);
    }
});

test('LINEARREG_SLOPE fast API in-place', () => {
    
    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
    const len = data.length;
    
    
    const ptr = wasm.linearreg_slope_alloc(len);
    
    try {
        
        const wasmData = new Float64Array(wasm.__wasm.memory.buffer, ptr, len);
        wasmData.set(data);
        
        
        wasm.linearreg_slope_into(ptr, ptr, len, 3);
        
        
        const result = new Float64Array(wasm.__wasm.memory.buffer, ptr, len);
        
        
        assert(isNaN(result[0]), 'Expected NaN at index 0');
        assert(isNaN(result[1]), 'Expected NaN at index 1');
        
        
        assert(!isNaN(result[2]), 'Expected valid value at index 2');
    } finally {
        
        wasm.linearreg_slope_free(ptr, len);
    }
});

test('LINEARREG_SLOPE batch API', () => {
    
    const close = new Float64Array(testData.close);
    
    const config = {
        period_range: [10, 20, 5]  
    };
    
    const result = wasm.linearreg_slope_batch(close, config);
    
    
    assert(result.values, 'Result should have values array');
    assert(result.combos, 'Result should have combos array');
    assert.strictEqual(result.rows, 3, 'Should have 3 rows');
    assert.strictEqual(result.cols, close.length, 'Columns should match input length');
    assert.strictEqual(result.values.length, 3 * close.length, 'Values array should be rows * cols');
    assert.strictEqual(result.combos.length, 3, 'Should have 3 parameter combinations');
    
    
    assert.strictEqual(result.combos[0].period, 10);
    assert.strictEqual(result.combos[1].period, 15);
    assert.strictEqual(result.combos[2].period, 20);
});

test('LINEARREG_SLOPE all NaN input', () => {
    
    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);
    
    assert.throws(() => {
        wasm.linearreg_slope_js(allNaN, 14);
    }, /All values are NaN/);
});

test('LINEARREG_SLOPE batch_into API', () => {
    
    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);
    const len = data.length;
    const periods = [5, 10]; 
    const rows = periods.length;
    const totalSize = rows * len;
    
    
    const inPtr = wasm.linearreg_slope_alloc(len);
    const outPtr = wasm.linearreg_slope_alloc(totalSize);
    
    try {
        
        const inView = new Float64Array(wasm.__wasm.memory.buffer, inPtr, len);
        inView.set(data);
        
        
        const resultRows = wasm.linearreg_slope_batch_into(
            inPtr,
            outPtr,
            len,
            5,  
            10, 
            5   
        );
        
        assert.strictEqual(resultRows, 2, 'Should return 2 rows');
        
        
        const results = new Float64Array(wasm.__wasm.memory.buffer, outPtr, totalSize);
        
        
        for (let i = 0; i < 4; i++) {
            assert(isNaN(results[i]), `Expected NaN at row 0, index ${i}`);
        }
        assert(!isNaN(results[4]), 'Expected valid value at row 0, index 4');
        
        
        const row2Start = len;
        for (let i = 0; i < 9; i++) {
            assert(isNaN(results[row2Start + i]), `Expected NaN at row 1, index ${i}`);
        }
        assert(!isNaN(results[row2Start + 9]), 'Expected valid value at row 1, index 9');
    } finally {
        
        wasm.linearreg_slope_free(inPtr, len);
        wasm.linearreg_slope_free(outPtr, totalSize);
    }
});

test('LINEARREG_SLOPE linear data', () => {
    
    const linearData = new Float64Array(20);
    for (let i = 0; i < 20; i++) {
        linearData[i] = 2 * i + 10;
    }
    
    const result = wasm.linearreg_slope_js(linearData, 14);
    
    
    for (let i = 13; i < result.length; i++) {
        assertClose(result[i], 2.0, 1e-9, 
                   `Expected slope=2.0 for linear data at index ${i}`);
    }
});

test('LINEARREG_SLOPE constant data', () => {
    
    const constantData = new Float64Array(20);
    constantData.fill(100.0);
    
    const result = wasm.linearreg_slope_js(constantData, 10);
    
    
    for (let i = 9; i < result.length; i++) {
        assertClose(result[i], 0.0, 1e-9, 
                   `Expected slope=0.0 for constant data at index ${i}`);
    }
});

test('LINEARREG_SLOPE batch with different warmup periods', () => {
    
    const close = new Float64Array(30);
    for (let i = 0; i < 30; i++) {
        close[i] = Math.sin(i * 0.1) * 100 + 1000;
    }
    
    const config = {
        period_range: [5, 15, 5]  
    };
    
    const result = wasm.linearreg_slope_batch(close, config);
    
    
    assert.strictEqual(result.rows, 3, 'Should have 3 rows');
    assert.strictEqual(result.cols, close.length, 'Columns should match input length');
    assert.strictEqual(result.combos.length, 3, 'Should have 3 parameter combinations');
    
    
    const periods = [5, 10, 15];
    for (let rowIdx = 0; rowIdx < periods.length; rowIdx++) {
        const period = periods[rowIdx];
        const rowStart = rowIdx * close.length;
        
        
        for (let i = 0; i < period - 1; i++) {
            assert(isNaN(result.values[rowStart + i]), 
                  `Expected NaN at row ${rowIdx}, index ${i} for period ${period}`);
        }
        
        
        for (let i = period - 1; i < close.length; i++) {
            assert(!isNaN(result.values[rowStart + i]), 
                  `Unexpected NaN at row ${rowIdx}, index ${i} for period ${period}`);
        }
    }
});

test.after(() => {
    console.log('LINEARREG_SLOPE WASM tests completed');
});

File: tests/python/test_linearreg_slope.py
"""
Python binding tests for LINEARREG_SLOPE indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path


sys.path.insert(0, str(Path(__file__).parent.parent.parent / 'target/wheels'))

try:
    import my_project as ta_indicators
except ImportError:
    
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS


class TestLinearRegSlope:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_linearreg_slope_partial_params(self, test_data):
        """Test LINEARREG_SLOPE with default parameters - mirrors check_linearreg_slope_partial_params"""
        close = test_data['close']
        
        
        result = ta_indicators.linearreg_slope(close, 14)
        assert len(result) == len(close)
    
    def test_linearreg_slope_accuracy(self):
        """Test LINEARREG_SLOPE matches expected values from Rust tests - mirrors check_linearreg_slope_accuracy"""
        input_data = np.array([100.0, 98.0, 95.0, 90.0, 85.0, 80.0, 78.0, 77.0, 79.0, 81.0])
        
        result = ta_indicators.linearreg_slope(input_data, period=5)
        
        assert len(result) == len(input_data)
        
        
        for i in range(4):
            assert np.isnan(result[i]), f"Expected NaN at index {i}, got {result[i]}"
        
        
        expected_values = [-3.8, -4.6, -4.4, -3.3, -1.5, 0.3]
        
        for i, expected in enumerate(expected_values, start=4):
            assert_close(result[i], expected, rtol=1e-9, atol=1e-9,
                        msg=f"LINEARREG_SLOPE value mismatch at index {i}")
    
    def test_linearreg_slope_zero_period(self):
        """Test LINEARREG_SLOPE fails with zero period - mirrors check_linearreg_slope_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.linearreg_slope(input_data, period=0)
    
    def test_linearreg_slope_period_exceeds_length(self):
        """Test LINEARREG_SLOPE fails when period exceeds data length - mirrors check_linearreg_slope_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.linearreg_slope(data_small, period=10)
    
    def test_linearreg_slope_very_small_dataset(self):
        """Test LINEARREG_SLOPE fails with insufficient data - mirrors check_linearreg_slope_very_small_dataset"""
        single_point = np.array([42.0])
        
        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.linearreg_slope(single_point, period=14)
    
    def test_linearreg_slope_empty_input(self):
        """Test LINEARREG_SLOPE fails with empty input"""
        empty = np.array([])
        
        with pytest.raises(ValueError, match="Empty data provided"):
            ta_indicators.linearreg_slope(empty, period=14)
    
    def test_linearreg_slope_nan_handling(self, test_data):
        """Test LINEARREG_SLOPE handles NaN values correctly - mirrors check_linearreg_slope_nan_handling"""
        close = test_data['close']
        
        result = ta_indicators.linearreg_slope(close, period=14)
        assert len(result) == len(close)
        
        
        if len(result) > 240:
            assert not np.any(np.isnan(result[240:])), "Found unexpected NaN after warmup period"
        
        
        assert np.all(np.isnan(result[:13])), "Expected NaN in warmup period"
    
    def test_linearreg_slope_all_nan_input(self):
        """Test LINEARREG_SLOPE with all NaN values"""
        all_nan = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.linearreg_slope(all_nan, period=14)
    
    def test_linearreg_slope_streaming(self, test_data):
        """Test LINEARREG_SLOPE streaming matches batch calculation - mirrors streaming tests"""
        close = test_data['close'][:100]  
        period = 14
        
        
        batch_result = ta_indicators.linearreg_slope(close, period=period)
        
        
        stream = ta_indicators.LinearRegSlopeStream(period=period)
        stream_values = []
        
        for price in close:
            result = stream.update(price)
            stream_values.append(result if result is not None else np.nan)
        
        stream_values = np.array(stream_values)
        
        
        assert len(batch_result) == len(stream_values)
        
        
        for i, (b, s) in enumerate(zip(batch_result, stream_values)):
            if np.isnan(b) and np.isnan(s):
                continue
            assert_close(b, s, rtol=1e-9, atol=1e-9, 
                        msg=f"LINEARREG_SLOPE streaming mismatch at index {i}")
    
    def test_linearreg_slope_batch_single(self, test_data):
        """Test LINEARREG_SLOPE batch processing with single parameter set"""
        close = test_data['close'][:100]  
        
        result = ta_indicators.linearreg_slope_batch(
            close,
            period_range=(14, 14, 0)  
        )
        
        assert 'values' in result
        assert 'periods' in result
        
        
        assert result['values'].shape[0] == 1
        assert result['values'].shape[1] == len(close)
        assert len(result['periods']) == 1
        assert result['periods'][0] == 14
        
        
        batch_row = result['values'][0]
        single_result = ta_indicators.linearreg_slope(close, period=14)
        
        
        for i in range(len(close)):
            if np.isnan(single_result[i]) and np.isnan(batch_row[i]):
                continue
            assert_close(batch_row[i], single_result[i], rtol=1e-9, atol=1e-9,
                        msg=f"Batch vs single mismatch at index {i}")
    
    def test_linearreg_slope_batch_multiple(self, test_data):
        """Test LINEARREG_SLOPE batch processing with multiple periods"""
        close = test_data['close'][:50]  
        
        result = ta_indicators.linearreg_slope_batch(
            close,
            period_range=(10, 20, 5)  
        )
        
        assert 'values' in result
        assert 'periods' in result
        
        
        assert result['values'].shape[0] == 3
        assert result['values'].shape[1] == len(close)
        assert len(result['periods']) == 3
        
        
        expected_periods = [10, 15, 20]
        for i, period in enumerate(expected_periods):
            assert result['periods'][i] == period
            
            
            batch_row = result['values'][i]
            single_result = ta_indicators.linearreg_slope(close, period=period)
            
            for j in range(len(close)):
                if np.isnan(single_result[j]) and np.isnan(batch_row[j]):
                    continue
                assert_close(batch_row[j], single_result[j], rtol=1e-9, atol=1e-9,
                            msg=f"Period {period} mismatch at index {j}")
    
    def test_linearreg_slope_linear_data(self):
        """Test LINEARREG_SLOPE with perfectly linear data"""
        
        linear_data = np.array([2*i + 10 for i in range(20)], dtype=np.float64)
        
        result = ta_indicators.linearreg_slope(linear_data, period=14)
        
        
        for i in range(13, len(result)):
            assert_close(result[i], 2.0, rtol=1e-9, atol=1e-9,
                        msg=f"Expected slope=2.0 for linear data at index {i}")
    
    def test_linearreg_slope_constant_data(self):
        """Test LINEARREG_SLOPE with constant data"""
        
        constant_data = np.full(20, 100.0)
        
        result = ta_indicators.linearreg_slope(constant_data, period=10)
        
        
        for i in range(9, len(result)):
            assert_close(result[i], 0.0, rtol=1e-9, atol=1e-9,
                        msg=f"Expected slope=0.0 for constant data at index {i}")
    
    def test_linearreg_slope_batch_warmup_periods(self, test_data):
        """Test that batch processing correctly handles different warmup periods"""
        close = test_data['close'][:30]
        
        result = ta_indicators.linearreg_slope_batch(
            close,
            period_range=(5, 15, 5)  
        )
        
        
        periods = [5, 10, 15]
        for row_idx, period in enumerate(periods):
            row = result['values'][row_idx]
            
            
            for i in range(period - 1):
                assert np.isnan(row[i]), f"Expected NaN at index {i} for period {period}"
            
            
            for i in range(period - 1, len(row)):
                assert not np.isnan(row[i]), f"Unexpected NaN at index {i} for period {period}"


if __name__ == '__main__':
    pytest.main([__file__, '-v'])
